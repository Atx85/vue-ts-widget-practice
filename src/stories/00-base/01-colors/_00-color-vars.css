:root {
    --product-widget-white: #FFF;
    --product-widget-black: #212121;
    --product-widget-blue: #2E3A8C;
    --product-widget-green: #3B755F;
    --product-widget-beige: #F2EBDB;
}
