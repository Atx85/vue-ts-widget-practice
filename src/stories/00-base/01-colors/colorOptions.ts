type ColorName = "white" | "black" | "blue" | "green" | "beige";
type Hex = `#${string}`;
type ColorsType = Record<ColorName, Hex>;
const colors: ColorsType = {
  white: "#FFF",
  black: "#212121",
  blue: "#2E3A8C",
  green: "#3B755F",
  beige: "#F2EBDB",
};
const ColorsTypeHexToColorName: Record<Hex, ColorName> = {
  "#FFF": "white",
  "#212121": "black",
  "#2E3A8C": "blue",
  "#3B755F": "green",
  "#F2EBDB": "beige",
};
type ColorMapType = Record<Hex, Hex>;
const secondaryColorMap: ColorMapType = {
  "#FFF": "#3B755F",
  "#212121": "#FFF",
  "#2E3A8C": "#FFF",
  "#3B755F": "#FFF",
  "#F2EBDB": "#3B755F",
};
export {
  ColorName,
  ColorsType,
  Hex,
  colors,
  secondaryColorMap,
  ColorsTypeHexToColorName,
};
