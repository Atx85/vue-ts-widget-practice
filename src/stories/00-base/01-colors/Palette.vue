<template>
  <ul class="palette">
    <li v-for="color in colors" :style="{ backgroundColor: color }">
      <p>{{ color }}</p>
    </li>
  </ul>
</template>
<style scoped>
.palette {
  list-style: none;
  display: flex;
  align-items: center;
  justify-content: center;
}
.palette li {
  width: 100px;
  height: 50px;
  border: 1px solid black;
  margin-right: 14px;
  list-style: none;
  display: flex;
  align-items: center;
  justify-content: center;
}
.palette li:last-of-type {
  margin-right: 0;
}
</style>

<script lang="ts">
import { defineComponent, PropType } from "vue";
import { colors, ColorsType } from "./colorOptions";

export default defineComponent({
  name: "Palette",
  props: {
    colors: {
      type: Object as PropType<ColorsType>,
      default: colors,
    },
  },
});
</script>
