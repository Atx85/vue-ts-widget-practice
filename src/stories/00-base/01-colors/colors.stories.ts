import { fn } from "@storybook/test";
import type { Meta, StoryObj } from "@storybook/vue3";

import Palette from "./Palette.vue";
import { ColorsType, colors } from "./colorOptions";

const meta: Meta<typeof Palette> = {
  title: "Base/Colors",
  component: Palette,
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Palette>;

export const ColorsOptions: Story = {
  args: {
    colors,
  },
};
