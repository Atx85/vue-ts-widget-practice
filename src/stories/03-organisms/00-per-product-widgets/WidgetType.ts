import { titleMapKey } from "@/stories/02-molecules/00-product-widget/TitleMapKey";
import { ActionType } from "./ActionType";
import { ColorName } from "@/stories/00-base/01-colors/colorOptions";

export interface WidgetType {
  id: number;
  type: titleMapKey;
  amount: number;
  action: ActionType;
  active: boolean;
  linked: boolean;
  selectedColor: ColorName;
}
