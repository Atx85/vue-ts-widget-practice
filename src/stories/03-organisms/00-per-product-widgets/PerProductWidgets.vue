<template>
  <div class="per-product-widgets">
    <h3 class="title font-primary-bold">{{ title }}</h3>
    <div class="widgets">
      <ProductWidget
        v-for="widget in localWidgets"
        :key="'widget' + widget.id"
        :id="(widget.id as number)"
        :type="widget.type"
        :amount="(widget.amount as number)"
        :action="widget.action"
        :active="widget.active"
        :linked="widget.linked"
        :selectedColor="colors[widget.selectedColor]"
      />
    </div>
  </div>
</template>

<style scoped>
.title {
  font-size: 30px;
  line-height: 36px;
  color: #212121;
  border-bottom: 2px solid #b0b0b0;
  margin-bottom: 20px;
  padding-bottom: 10px;
}
.per-product-widgets {
  padding: 36px;
  display: inline-block;
  background-color: #f9f9f9;
  box-shadow: 0px 0.78px 2.99px 0px #00000003, 0px 3.42px 6.19px 0px #00000005,
    0px 8.4px 12.35px 0px #00000006, 0px 16.18px 24.21px 0px #00000008,
    0px 27.22px 44.51px 0px #0000000a, 0px 42px 76px 0px #0000000d;
  border-radius: 8px;
  max-width: calc(100% - 40px);
  width: 300px;
}
.widgets {
  display: flex;
  flex-flow: column;
  width: 100%;
}
.product-widget {
  margin-bottom: 36.95px;
}
@media (min-width: 600px) {
  .widgets {
    display: grid;
    grid-template-columns: repeat(2, auto);
    min-width: 500px;
  }
  .per-product-widgets {
    width: calc(100% - 40px);
    max-width: calc(100% - 40px);
    max-width: 640px;
  }
}

@media (min-width: 900px) {
  .widgets {
    display: grid;
    grid-template-columns: repeat(3, auto);
    column-gap: 57.52px;
    width: 100%;
    row-gap: 0;
  }
  .per-product-widgets {
    max-width: 851px;
  }
}
</style>

<script lang="ts">
import { defineComponent, PropType } from "vue";
import "../../00-base/00-defaults/00-defaults.css";
import ProductWidget from "../../02-molecules/00-product-widget/ProductWidget.vue";
import {
  colors,
  ColorName,
  Hex,
  ColorsTypeHexToColorName,
} from "../../00-base/01-colors/colorOptions";
import { titleMapKey } from "../../02-molecules/00-product-widget/TitleMapKey";
import { ActionType } from "./ActionType";
import { WidgetType } from "./WidgetType";

interface WidgetUpdateInfo {
  id: number;
  active: boolean;
  linked: boolean;
  selectedColor: Hex;
}

export default defineComponent({
  name: "per-product-widget",
  components: { ProductWidget },
  props: {
    widgets: {
      type: Array as PropType<Array<WidgetType>>,
      default: () => [],
    },
    title: String,
  },
  computed: {
    localWidgets(): Array<WidgetType> {
      return this.widgets;
    },
  },
  data() {
    return {
      colors,
      titleMap: {
        carbon: "kgs of carbon",
        "plastic bottles": " plastic bottles",
        trees: " trees",
      },
    };
  },
});
</script>
