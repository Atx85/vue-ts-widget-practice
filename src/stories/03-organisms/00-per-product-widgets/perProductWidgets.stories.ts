import PerProductWidgets from "./PerProductWidgets.vue";
import type { Meta, StoryObj } from "@storybook/vue3";
import { colors } from "../../00-base/01-colors/colorOptions";

const meta: Meta<typeof PerProductWidgets> = {
  title: "Organisms/Per Product Widget ",
  component: PerProductWidgets,
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof PerProductWidgets>;

export const PerProductWidgetsOptions: Story = {
  args: {
    widgets: [
      {
        id: 1,
        type: "carbon",
        amount: 2,
        action: "offsets",
        active: false,
        linked: false,
        selectedColor: "green",
      },
      {
        id: 2,
        type: "trees",
        amount: 15,
        action: "plants",
        active: false,
        linked: true,
        selectedColor: "black",
      },
      {
        id: 3,
        type: "plastic",
        amount: 300,
        action: "collects",
        active: true,
        linked: true,
        selectedColor: "beige",
      },
    ],
    title: "Per product widgets",
  },
};
