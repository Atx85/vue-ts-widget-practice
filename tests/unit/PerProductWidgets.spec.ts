import { mount } from "@vue/test-utils";
import PerProductWidgets from "@/stories/03-organisms/00-per-product-widgets/PerProductWidgets.vue";
import { WidgetType } from "@/stories/03-organisms/00-per-product-widgets/WidgetType";
import { ColorName, colors } from "@/stories/00-base/01-colors/colorOptions";
describe("PerProductWidgets.vue", () => {
  const widgetsData = [
    {
      id: 1,
      type: "carbon",
      amount: 2,
      action: "offsets",
      active: false,
      linked: false,
      selectedColor: "green",
    },
    {
      id: 2,
      type: "trees",
      amount: 15,
      action: "plants",
      active: false,
      linked: true,
      selectedColor: "black",
    },
    {
      id: 3,
      type: "plastic",
      amount: 300,
      action: "collects",
      active: true,
      linked: true,
      selectedColor: "beige",
    },
  ];
  const wrapper = mount(PerProductWidgets, {
    props: {
      widgets: widgetsData as Array<WidgetType>,
      title: "Per product widgets",
    },
  });
  it("should render", () => {
    expect(wrapper.find("h3.title").text()).toContain("Per product widgets");
    expect(wrapper.findAll(".product-widget")).toHaveLength(3);
  });

  it("should have the 'Linked to Public Profile' checked based on config", () => {
    const widgets = wrapper.findAll(".product-widget");
    expect(widgets[0].find("#linkToProfileid_1").attributes().value).toBe(
      widgetsData[0].linked ? "true" : "false"
    );
    expect(widgets[1].find("#linkToProfileid_2").attributes().value).toBe(
      widgetsData[1].linked ? "true" : "false"
    );
    expect(widgets[2].find("#linkToProfileid_3").attributes().value).toBe(
      widgetsData[2].linked ? "true" : "false"
    );
  });

  it("should have the 'Activate badge' checked based on config", () => {
    const widgets = wrapper.findAll(".product-widget");
    expect(widgets[0].find("#activateBadgeid_1").attributes().value).toBe(
      widgetsData[0].active ? "true" : "false"
    );
    expect(widgets[1].find("#activateBadgeid_2").attributes().value).toBe(
      widgetsData[1].active ? "true" : "false"
    );
    expect(widgets[2].find("#activateBadgeid_3").attributes().value).toBe(
      widgetsData[2].active ? "true" : "false"
    );
  });
  it("should have the selected colour based on config", () => {
    const widgets = wrapper.findAll(".product-widget");
    expect(
      widgets[0]
        .find(".colour-box-view.active + input[type=checkbox]")
        .attributes().value
    ).toBe(colors[widgetsData[0].selectedColor as ColorName]);
    expect(
      widgets[1]
        .find(".colour-box-view.active + input[type=checkbox]")
        .attributes().value
    ).toBe(colors[widgetsData[1].selectedColor as ColorName]);
    expect(
      widgets[2]
        .find(".colour-box-view.active + input[type=checkbox]")
        .attributes().value
    ).toBe(colors[widgetsData[2].selectedColor as ColorName]);
  });
  it("should have the 'Activate badge' checked based on config", () => {
    const widgets = wrapper.findAll(".product-widget");
    expect(widgets[0].find("#activateBadgeid_1").attributes().value).toBe(
      "false"
    );
    expect(widgets[1].find("#activateBadgeid_2").attributes().value).toBe(
      "false"
    );
    expect(widgets[2].find("#activateBadgeid_3").attributes().value).toBe(
      "true"
    );
  });
});
