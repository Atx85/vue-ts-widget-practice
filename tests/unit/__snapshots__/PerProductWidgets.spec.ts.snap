// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerProductWidgets.vue should render 1`] = `
<div class="per-product-widgets">
  <h3 class="title font-primary-bold"></h3>
  <div class="widgets"></div>
</div>
`;
